crates/adapters/src/transport/durable/kafka/input.rs

```diff
@@ -0,0 +1,771 @@
+    ///   the data topic.
+    fn find_step(&self, step: Step) -> AnyResult<(i64, i64)> {
+        let ctp = Ctp::new(&self.consumer, self.topic, self.partition);
+        let mut events = self.watermarks.clone();
+        loop {
+            assert!(!events.is_empty());
+            let offset = events.start + (events.end - events.start) / 2;
+            let entry = IndexEntry::read(&ctp, offset)?;
+            events = match step.cmp(&entry.step) {
+                Ordering::Equal => {
+                    ctp.pause()?;
+                    return Ok((offset, entry.data_offsets.start));
+                }
+                Ordering::Greater => offset + 1..events.end,
+                Ordering::Less => events.start..offset,
+            };
+        }
+    }
+}
+
+impl Endpoint {
+    fn new(config: KafkaDurableInputConfig) -> AnyResult<Self> {
+        Ok(Self(config.try_into()?))
+    }
+}
+
+fn poll_consumer<C: ConsumerContext>(consumer: &BaseConsumer<C>) -> KafkaResult<()> {
+    match consumer.poll(Duration::ZERO) {
+        None => Ok(()),
+        Some(Ok(_)) => {
+            // Should not be possible because we split off partition queues for
+            // all the partitions we subscribe.
+            unreachable!()
+        }
+        Some(Err(error)) => Err(error),
+    }
+}
+
+fn split_partition_queue<C: ConsumerContext>(
+    consumer: &Arc<BaseConsumer<C>>,
+    topic: &str,
+    partition: i32,
+) -> AnyResult<PartitionQueue<C>> {
+    consumer
+        .split_partition_queue(topic, partition)
+        .ok_or_else(|| anyhow!("invalid topic {topic} or partition {partition}"))
+}
+
+fn read_partition_queue<'a, C: ConsumerContext>(
+    consumer: &BaseConsumer<C>,
+    partition: &'a PartitionQueue<C>,
+) -> KafkaResult<BorrowedMessage<'a>> {
+    loop {
+        poll_consumer(consumer)?;
+        if let Some(result) = partition.poll(POLL_TIMEOUT) {
+            return result;
+        }
+    }
+}
+
+struct KafkaInputContext {
+    sender: Sender<(Step, AnyResult<()>)>,
+    start: Instant,
+}
+
+impl KafkaInputContext {
+    fn new(sender: Sender<(Step, AnyResult<()>)>) -> Self {
+        Self {
+            sender,
+            start: Instant::now(),
+        }
+    }
+}
+
+impl ClientContext for KafkaInputContext {
+    fn error(&self, _error: KafkaError, _reason: &str) {
+        // eprintln!("Kafka error: {error}");
+        /*
+                if let Some(endpoint) = self.endpoint.lock().unwrap().upgrade() {
+                    endpoint.push_error(error, reason);
+                }
+        */
+    }
+
+    fn log(&self, level: RDKafkaLogLevel, fac: &str, log_message: &str) {
+        println!(
+            "log: {}ms {level:?} {fac} {log_message}",
+            self.start.elapsed().as_millis()
+        );
+    }
+
+    fn stats(&self, statistics: rdkafka::Statistics) {
+        println!("stats: {:?}", statistics)
+    }
+}
+
+impl ProducerContext for KafkaInputContext {
```

Comment:
> I'm a little confused: why is it called `InputContext` if it's a `ProducerContext`? Shouldn't this be an `OutputContext` then?

It's because the whole file is about input.

It's a confusing name. I'll change it.